openhands/runtime/impl/action_execution/action_execution_client.py

```diff
@@ -264,6 +265,12 @@ def send_action_for_execution(self, action: Action) -> Observation:
                 raise AgentRuntimeTimeoutError(
                     f'Runtime failed to return execute_action before the requested timeout of {action.timeout}s'
                 )
+            except requests.HTTPError as e:
+                if e.response.status_code in (404, 502, 503):
+                    raise AgentRuntimeNotFoundError(
+                        'Runtime unavailable: System resources may be exhausted due to running commands. This may be fixed by retrying.'
+                    ) from e
+                raise e
```

Comment:
I'd suggest putting this in the `remote` implementation of `_send_request`

Reply:
done!